kotlinx-coroutines-core/common/src/flow/terminal/Reduce.kt

```diff
@@ -120,3 +120,39 @@ public suspend fun <T> Flow<T>.first(predicate: suspend (T) -> Boolean): T {
     if (result === NULL) throw NoSuchElementException("Expected at least one element matching the predicate $predicate")
     return result as T
 }
+
+/**
+ * The terminal operator that returns the first element emitted by the flow and then cancels flow's collection.
+ * Returns [null] if the flow was empty.
+ */
+public suspend fun <T> Flow<T>.firstOrNull(): T? {
```

Comment:
Looking at `singleOrNull()`, it takes in a `T : Any`. Wouldn't we want to do the same as well for `firstOrNull()`? Otherwise, given a `Flow<Foo?>`, there would be ambiguity between a `null` emission and a completion without any emissions.

Reply:
I was wondering about that. It would be ambiguous if you needed to distinguish between a null element and a null fallback, but it comes down to semantics on whether you actually care. If we look at the [`collections`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/first-or-null.html) and [`sequences`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.sequences/first-or-null.html) APIs, there is no `T : Any` requirement since there may be a totally valid reason you don't care how you got `null`.

To be consistent, I'm also fine with `T : Any` purely for consistency with existing `Flow` APIs.

Reply:
We've discussed it today and had decided to leave `xxxOrNull` on Kotlin Flow as extensions to `<T : Any>`. Note, that this is not consistent with stdlib (where you can use `Any?` in similar operators) and it was made so because of pre-v1.0 reports that it is not convenient. We don't know if the same would apply to Flow, so leaving it as `<T : Any>` for now is the conservative decision here. If we find use-cases where it causes inconvenience it is always safe to wide the upper bound to `Any?` in future releases.

Reply:
Perfect, I'll apply that update now